database/src/error.rs

```diff
@@ -2,61 +2,148 @@
-    /// TODO
-    #[error("database PATH is inaccessible: {0}")]
-    Path(std::io::Error),
+pub enum InitError {
+    /// I/O error.
+    #[error("database I/O error: {0}")]
+    Io(#[from] std::io::Error),
+
+    /// The given `Path/File` existed and was accessible,
+    /// but was not a valid database file.
+    #[error("database file exists but is not valid")]
+    Invalid,
+
+    /// The given `Path/File` existed, was a valid
+    /// database, but the version is incorrect.
+    #[error("database file is valid, but version is incorrect")]
+    InvalidVersion,
 
-    /// TODO
-    #[error("{DATABASE_BACKEND} error: {0}")]
-    Backend(BackendError),
+    /// The given `Path/File` existed,
+    /// was a valid database, but it is corrupt.
+    #[error("database file is corrupt")]
+    Corrupt,
 
-    /// TODO
+    /// The database is currently in the process
+    /// of shutting down and cannot respond.
+    ///
+    /// TODO: This might happen if we try to open
+    /// while we are shutting down, `unreachable!()`?
+    #[error("database is shutting down")]
+    ShuttingDown,
+
+    /// A fatal error occurred.
+    ///
+    /// This is for errors that _should_ be unreachable
+    /// but we'd still like to panic gracefully.
     ///
+    /// # Invariant
+    /// If this error is received, all(?) of Cuprate should shutdown.
+    #[error("fatal error: {0}")]
+    Fatal(BoxError),
+
     /// An unknown error occurred.
+    ///
+    /// This is a catch-all for all non-fatal errors.
     #[error("unknown error: {0}")]
-    Unknown(Cow<'static, str>),
+    Unknown(BoxError),
 }
 
 //---------------------------------------------------------------------------------------------------- RuntimeError
-/// Database errors that occur _after_ successful initialization.
-#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
-#[cfg_attr(
-    feature = "borsh",
-    derive(borsh::BorshSerialize, borsh::BorshDeserialize)
-)]
-#[derive(thiserror::Error, Debug, Clone, PartialEq, PartialOrd, Eq, Ord, Hash)]
+/// Errors that occur _after_ successful ([`Env::open`]).
+///
+/// There are no errors for:
+/// 1. Missing tables
+/// 2. (De)serialization
+///
+/// as `cuprate_database` upholds the invariant that:
+///
+/// 1. All tables exist
+/// 2. (De)serialization never fails
+#[derive(thiserror::Error, Debug)]
 pub enum RuntimeError {
-    // TODO: replace string with actual error type.
+    /// The given key already existed in the database.
+    #[error("key already existed")]
+    KeyExists,
+
+    /// The given key did not exist in the database.
+    #[error("key/value pair was not found")]
+    KeyNotFound,
+
+    /// The database environment has reached
+    /// maximum memory map size, it must be
+    /// increased.
+    //
+    // TODO: `sanakirja` automatically resizes, `heed` does not.
+    // I guess this should be `unreachable!()` for `sanakirja`?
+    #[error("not enough space in database environment memory map")]
+    MapFull,
```

Comment:
We do need to handle this for `heed` as well, in the DB. 

As LMDB according to the docs here: https://docs.rs/heed/0.20.0-alpha.9/heed/struct.Env.html#method.resize only allows resizing when there are no other txs, we need a way for a RWTX to pause reader and for it to check that all readers are paused before resizing.

We would probably need a static `RWLock`, all reader will acquire a read lock and hold the guard for the duration of the tx, to do this we would have to create a wrapper type for heeds tx which includes the guard. Then heed writters will only obtain the write lock when they need to resize the env. This would be a little bit of overhead but I can't think of a better way

Reply:
Will implement next PR. We need some type of signal for `shutdown` too, the writer thread will be busy flushing to disk & hanging on [`Env::prepare_for_closing`](https://docs.rs/heed/0.20.0-alpha.9/heed/struct.Env.html#method.prepare_for_closing) so it won't be able to respond with `RuntimeError::ShuttingDown`.

Instead of `RuntimeError::MapFull` existing, the writer thread should just automatically resize when needed, right? From the perspective of the writer thread:

1. Receive write request
2. Execute transaction
3. If transaction returns resize error, resize and try again until success
4. Respond

Reply:
What code needs to be ran on shutdown? can't we just leave it for heed and sanakirja's drop impl?

> Instead of RuntimeError::MapFull existing, the writer thread should just automatically resize when needed, right? From the perspective of the writer thread:

Yeah, step 3 should probably be bounded so we don't have an infinite resize issue if something goes wrong in the DB but that is the idea

Reply:
> We need some type of signal for shutdown too, the writer thread will be busy flushing to disk & hanging on [Env::prepare_for_closing](https://docs.rs/heed/0.20.0-alpha.9/heed/struct.Env.html#method.prepare_for_closing) so it won't be able to respond with RuntimeError::ShuttingDown.

iirc, reth implementation use a simple `Arc<AtomicBool>` in `Relaxed` ordering. This is sufficient imo, all reader threads drop out their env reference on Shutdown signal, and the write thread will eventually be freed.

Reply:
> What code needs to be ran on shutdown?

Just signaling that we're shutting down and flushing to disk I think? Both [`heed`](https://docs.rs/heed/0.20.0-alpha.9/src/heed/env.rs.html#321) and [`sanakirja`](https://docs.rs/sanakirja/latest/src/sanakirja/environment/mod.rs.html#73) don't seem to flush in their drop impls.

> iirc, reth implementation use a simple Arc\<AtomicBool\> in Relaxed ordering

They use `SeqCst`. It needs to be at least `Acquire+Release` or else readers may continue even after the writer has signaled it is shutting down. This does mean readers will `load()` an atomic bool each read, probably negligible in the grand scheme though.

Reply:
> Just signaling that we're shutting down and flushing to disk I think?

Maybe have a behavior so that while shutdown the tower::Service don't return poll_ready. But even then the database is the only primary components that need graceful shutdown. The rest could shutdown abruptly without any consequences.

> This does mean readers will load() an atomic bool each read, probably negligible in the grand scheme though.

This would be the cheapest way to do it. Using a channel would internally use an AtomicPtr anyway.

Reply:
Shutdown should only happen when both DB handles are dropped so we could just add a `shutdown` method to `Env` and impl `Drop` for `Env` just calling that:

```rust
impl<T: Env> Drop for T {
   fn drop(&self) {
      if !self.shutdown() {
          println!("Failed to shutdown the DB cleanly");
      }
   }
}
```

Reply:
^that assumes `shutdown` returns a bool, it could return a Result

Reply:
but txs, when they are committed are flushed to disk, right? so there should be no situation where the DB is dropped without a tx getting submitted if we shutdown when all handles are dropped.

So I don't think it's necessary us doing custom shutdown logic unless we were to allow a caller to shutdown the DB without all handles being dropped (which I would rather not have)

Reply:
> but txs, when they are committed are flushed to disk, right?

Yes I think by default [LMDB flushes to disk per transaction commit](http://www.lmdb.tech/doc/group__mdb.html#ga85e61f05aa68b520cc6c3b981dba5037) but it'll be the slowest option, doubly bad for HDD. It would be equivalent to [`monerod --db-sync-mode=safe:sync`](https://monero.stackexchange.com/questions/8575/more-detailed-explanations-for-the-db-sync-mode-parameters-i-e-safe-fast-fa).

If we wanted something like `monerod --db-sync-mode=fastest:async` then we would use `MDB_NOSYNC` and only `sync()` at the very end.

We can also support both and everything in-between like `monerod`.

Reply:
`sanakirja` seems to flush per transaction commit too https://docs.rs/sanakirja/latest/src/sanakirja/environment/muttxn.rs.html#330-333.

Reply:
ah ok then yeah, custom shutdown code it is